java-checks/src/main/java/org/sonar/java/checks/VolatileVariablesIncrementDecrementCheck.java

```diff
@@ -0,0 +1,97 @@
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+
+@Rule(key = "S3078")
+public class VolatileVariablesIncrementDecrementCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE_TEMPLATE = "Use an \"%s\" for this field; its %s are atomic.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.PREFIX_INCREMENT, Tree.Kind.POSTFIX_INCREMENT, Tree.Kind.PREFIX_DECREMENT, Tree.Kind.POSTFIX_DECREMENT, Tree.Kind.LOGICAL_COMPLEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    UnaryExpressionTree unaryExpression = (UnaryExpressionTree) tree;
+    ExpressionTree expression = ExpressionUtils.skipParentheses(unaryExpression.expression());
+    IdentifierTree identifier = getVariableIdentifier(expression);
+
+    if (identifier == null || !identifier.symbol().isVolatile()) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.LOGICAL_COMPLEMENT)) {
+      checkBooleanToggling(tree, identifier);
+    } else if (tree.is(Tree.Kind.PREFIX_INCREMENT) || tree.is(Tree.Kind.POSTFIX_INCREMENT)) {
+      checkIncrementDecrement(tree, identifier, "increments");
+    } else {
+      checkIncrementDecrement(tree, identifier, "decrements");
+    }
+  }
+
+  @Nullable
+  private static IdentifierTree getVariableIdentifier(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return (IdentifierTree) expressionTree;
+    } else if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      return ((MemberSelectExpressionTree) expressionTree).identifier();
+    }
+    return null;
+  }
+
+  private void checkBooleanToggling(@Nullable Tree tree, IdentifierTree identifier) {
+    Tree parent = tree != null ? tree.parent() : null;
+    if (parent != null) {
+      if (parent.is(Tree.Kind.PARENTHESIZED_EXPRESSION)) {
+        checkBooleanToggling(parent, identifier);
+      } else if (parent.is(Tree.Kind.ASSIGNMENT)) {
+        ExpressionTree variable = ((AssignmentExpressionTree) parent).variable();
+        if ((variable.is(Tree.Kind.IDENTIFIER) && identifier.name().equals(((IdentifierTree) variable).name()))
+          || (variable.is(Tree.Kind.MEMBER_SELECT) && identifier.name().equals(((MemberSelectExpressionTree) variable).identifier().name()))) {
+          reportIssue(parent, "Use an \"AtomicBoolean\" for this field");
+        }
+      }
+    }
+  }
+
+  private void checkIncrementDecrement(Tree tree, IdentifierTree identifier, String type) {
+    if (identifier.symbol().type().is("int") || identifier.symbol().type().is("java.lang.Integer")) {
```

Comment:
Use a dedicated variable instead of calling 4 times `identifier.symbol().type()`